scanner/e2etests/image_test.go

```diff
@@ -0,0 +1,244 @@
+		return fmt.Sprintf("%s:%s", d.GetDid(), d.GetVersionId())
+	}
+	return "unknown"
+}
+
+func (tc *TestCase) authConfig() authn.Authenticator {
+	return authn.FromConfig(authn.AuthConfig{
+		Username: os.Getenv(tc.Username),
+		Password: os.Getenv(tc.Password),
+	})
+}
+
+func (tc *TestCase) convertReport(vr *v4.VulnerabilityReport) *TestWant {
+	return &TestWant{
+		Namespace: namespace(vr),
+		Source:    tc.Source,
+		Features:  tc.convertFeatures(vr),
+	}
+}
+
+func (tc *TestCase) convertFeatures(vr *v4.VulnerabilityReport) []Feature {
+	nv := func(n, v string) string {
+		return fmt.Sprintf("%s-%s", n, v)
+	}
+	// Populate map with all features in the test case.
+	feats := make(map[string]*Feature, len(tc.Features))
+	for _, f := range tc.Features {
+		feats[nv(f.Name, f.Version)] = &f
+	}
+	// Convert every expected package in the report.
+	ret := make([]Feature, 0, len(tc.Features))
+	for _, p := range vr.GetContents().GetPackages() {
+		f, ok := feats[nv(p.Name, p.Version)]
+		if !ok {
+			// We only care for features in the test case.
+			continue
+		}
+		ret = append(ret, Feature{
+			Name:            p.GetName(),
+			NamespaceName:   namespace(vr),
+			Version:         p.GetVersion(),
+			Vulnerabilities: tc.convertVulns(vr, p, f),
+			// TODO Pending fields not currently available in the vulnerability report.
+			VersionFormat: f.VersionFormat,
+			AddedBy:       f.AddedBy,
+			Location:      f.Location,
+			FixedBy:       f.FixedBy,
+		})
+	}
+	return ret
+}
+
+func (tc *TestCase) convertVulns(vr *v4.VulnerabilityReport, pkg *v4.Package, feat *Feature) []Vulnerability {
+	// Populate map with all vulnerabilities in the test case feature.
+	featVulns := make(map[string]*Vulnerability)
+	for _, featVuln := range feat.Vulnerabilities {
+		featVulns[featVuln.Name] = &featVuln
+	}
+	// Convert all package vulnerabilities.
+	var vulns []Vulnerability
+	for _, id := range vr.GetPackageVulnerabilities()[pkg.GetId()].GetValues() {
+		v := vr.GetVulnerabilities()[id]
+		n, _, _ := strings.Cut(v.GetName(), " ")
+		featVuln, ok := featVulns[n]
+		if !ok && tc.OnlyCheckSpecifiedVulns {
+			// Skip since we only check what is in the test features.
+			continue
+		}
+		// Link is by default the first on the list, but we check if the expected link is
+		// somewhere else.
+		link, _, _ := strings.Cut(v.GetLink(), " ")
+		for _, l := range strings.Split(v.GetLink(), " ") {
+			if l == featVuln.Link {
+				link = l
+			}
+		}
+		// Convert severity.
+		severity, ok := map[v4.VulnerabilityReport_Vulnerability_Severity]string{
+			v4.VulnerabilityReport_Vulnerability_SEVERITY_CRITICAL:    "Critical",
+			v4.VulnerabilityReport_Vulnerability_SEVERITY_IMPORTANT:   "Important",
+			v4.VulnerabilityReport_Vulnerability_SEVERITY_MODERATE:    "Moderate",
+			v4.VulnerabilityReport_Vulnerability_SEVERITY_LOW:         "Low",
+			v4.VulnerabilityReport_Vulnerability_SEVERITY_UNSPECIFIED: "Unspecified",
+		}[v.GetNormalizedSeverity()]
+		if !ok {
```

Comment:
This is just for safety, right? We don't expect to reach here except for potential programmer error? If so, can we add an error log (or make the test fail) so the programmer knows to fix the underlying issue?

Reply:
There will be a logging failure since the test severity does not match the expected severity, same with "namespace". Error logging will be hard to inspect among all the other messages when the failure happens, so I am not sure how useful it is compared to seeing the assertion failure.